lib/morpher/transform.rb

```diff
@@ -0,0 +1,447 @@
+      #
+      # @param [Object] input
+      #
+      # @return [Either<Error, Array<Object>>]
+      def call(input)
+        PRIMITIVE
+          .call(input)
+          .lmap(&method(:lift_error))
+          .bind(&method(:run))
+      end
+
+    private
+
+      # rubocop:disable Metrics/MethodLength
+      def run(input)
+        output = []
+
+        input.each_with_index do |value, index|
+          output << transform.call(value).lmap do |error|
+            return failure(
+              error(
+                cause:   Index.wrap(error, index),
+                message: MESSAGE % { index: index },
+                input:   input
+              )
+            )
+          end.from_right
+        end
+
+        success(output)
+      end
+      # rubocop:enable Metrics/MethodLength
+    end # Array
+
+    # Transform a hash via mapping it over key specific transforms
+    class Hash < self
+      include Anima.new(:optional, :required)
+
+      KEY_MESSAGE = 'Missing keys: %<missing>s, Unexpected keys: %<unexpected>s'
+      PRIMITIVE   = Primitive.new(::Hash)
+
+      private_constant(*constants(false))
+
+      # Transform to symbolize array keys
+      class Symbolize < Transform
+        # Apply transformation to input
+        #
+        # @param [Hash{String => Object}]
+        #
+        # @return [Hash{Symbol => Object}]
+        def call(input)
+          success(input.transform_keys(&:to_sym))
+        end
+      end # Symbolize
+
+      # Key specific transformation
+      class Key < Transform
+        include Concord::Public.new(:value, :transform)
+
+        # Rendering slug
+        #
+        # @return [String]
+        def slug
+          '[%<key>s]' % { key: value.inspect }
+        end
+        memoize :slug
+
+        # Apply transformation to input
+        #
+        # @param [Object]
+        #
+        # @return [Either<Error, Object>]
+        def call(input)
+          transform.call(input).lmap do |error|
+            error(cause: error, input: input)
+          end
+        end
+      end # Key
+
+      # Apply transformation to input
+      #
+      # @param [Object] input
+      #
+      # @return [Either<Error, Object>]
+      def call(input)
+        PRIMITIVE
+          .call(input)
+          .lmap(&method(:lift_error))
+          .bind(&method(:reject_keys))
+          .bind(&method(:transform))
+      end
+
+    private
+
+      def transform(input)
+        transform_required(input).bind do |required|
+          transform_optional(input).fmap(&required.public_method(:merge))
+        end
+      end
+
+      def transform_required(input)
+        transform_keys(required, input)
+      end
+
+      def transform_optional(input)
+        transform_keys(
+          optional.select { |key| input.key?(key.value) },
+          input
+        )
+      end
+
+      # rubocop:disable Metrics/MethodLength
+      def transform_keys(keys, input)
+        success(
+          keys
+            .map do |key|
```

Comment:
Enumerable#to_h takes a block now. You can replace map here with it and remove the trailing to_h.